improve readme and add contributing file (#122)
cloudflare/next-on-pages: README.md

> # ⚡▲ next-on-pages ▲⚡

Reviewer: I think continuing to refer to this package as `@cloudflare/next-on-pages` rather than `next-on-pages` will have the advantage of reducing confusion when people try to install this. Wrangler v1 had a recurring problem of people trying to install `wrangler` rather than `@cloudflare/wrangler`.

Reviewer: I'd suggest changing it everywhere in this doc.

Author: ah ok 😓, I'll update this and the rest of the docs 👍

Author: Updated all throughout the docs 🙂👍 

I've also noticed this one:
https://github.com/cloudflare/next-on-pages/blob/cc91567c460e6fdc6915448245876bb5744f34d9/src/index.ts#L569

for clarity I guess we'd also. want to say `@cloudflare/next-on-pages` there as well right?
anyways if ok I'd leave it for another PR as I am trying to keep this documentation only 🙂:+1:

> next-on-pages is a CLI tool that you can use to build and develop [Next.js](https://nextjs.org/) applications so that they can run on the [Cloudflare Pages](https://pages.cloudflare.com/) platform (and integrate with the various Cloudflare offerings such as KV stores, and Durable Objects).

Reviewer: ```suggestion
next-on-pages is a CLI tool that you can use to build and develop [Next.js](https://nextjs.org/) applications so that they can run on the [Cloudflare Pages](https://pages.cloudflare.com/) platform (and integrate with Cloudflare's various other product offerings such as KV, D1, R2 and Durable Objects).
```

> This tool is a best-effort library implemented by the Cloudflare team and the community since Vercel doesn't and likely won't support Next.js running on Cloudflare officially. As such most but not all Next.js features are yet supported, see the [Supported Versions and Features document](./docs/supported.md) for more details.

Reviewer: ```suggestion
This tool is a best-effort library implemented by the Cloudflare team and the community. As such, most, but not all, Next.js features are supported. See the [Supported Versions and Features document](./docs/supported.md) for more details.
```

## Quick Start

This section describes how to bundle and deploy a (new or existing) Next.js application and using next-on-pages.

> > Note: this readme uses `npm`, but nothing relies on it, you can instead use `yarn` or `pnpm` if you want.

Author: always mentioning `npm or yarn or pnpm` is a bit too verbose to me, I think it's much cleaner just to stick with `npm` and tell the dev that they can use the others if they want (converting from one to the other should be trivial anyways)

Author: I might need to rethink this statement, as using `pnpm` doesn't seem to work out of the box (see the last comments here: https://github.com/cloudflare/next-on-pages/issues/52)

Author: I've removed the note for now since `pnpm` doesn't work out of the box, we can leave it as is for now and add some info later if we want

### 1. Create Next App

> To start creating a Next.js application using next-on-pages, start by creating a standard Next.js application using the create-next-app command (skip this step if you already have an existing Next.js application that you want to port to Cloudflare using next-on-pages):

Reviewer: ```suggestion
To start using `@cloudflare/next-on-pages`, you must first have a Next.js project you wish to deploy. If you don't already have a project, you can use the `create-next-app` command:
```

```sh

> npx create-next-app@13.2.4 my-next-app

Reviewer: ```suggestion
npx create-next-app@latest
```

Is there any reason to specify the version?

Author: yeah as you saw in https://github.com/cloudflare/next-on-pages/pull/122/files#r1147455407

`@latest` feels a bit risky, I don't mind keeping it for now and we can decide later if we want to revisit it 🙂

1. `cd` into the new directory (e.g. `cd my-app`)
### 1. Create Next App

1. `npm install -D @cloudflare/next-on-pages vercel`
To start creating a Next.js application using next-on-pages, start by creating a standard Next.js application using the create-next-app command (skip this step if you already have an existing Next.js application that you want to port to Cloudflare using next-on-pages):

1. Configure the project to use the Edge Runtime:
```sh
npx create-next-app@13.2.4 my-next-app
```

>    Replace `pages/api/hello.ts` with the following:

Author: I don't think we can really tell the developer what to change since the create-next-app offers the option to use `src`, `app` or neither, so I changes the message to be less specific (I think that adding details for all possible choices would be way too verbose)

> The above command suggests to use Next's `13.2.4` version, that is the latest at the time of writing, older versions are also supported (but they might be only partially), newer versions might not be so be wary of that. The team will soon test various versions and document more clearly which are supported to what degree.

Reviewer: Eh, I'd probably just let them use `@latest`.

```suggestion
We have confirmed support for to the current version of Next.js, at the time of writing, `13.2.4`. Although we'll endeavor to keep support for newer versions, we cannot guarantee that we'll always be up-to-date with the latest version. If you experience any problems with `@cloudflare/next-on-pages`, you may wish to try pinning to `13.2.4` while we work on supporting any recent breaking changes.
```

Change your current directory to the newly created one, as the following step require you to be in it:

```sh
cd my-next-app
```

### 2. Configure the application to use the Edge Runtime

> In order for the application to run on Cloudflare Pages it needs to be set to run on the edge runtime, in order to do so make sure that all the files in your application containing server side code (as your api routes and pages using `getServerSideProps`) export a config object specifying the use of the edge runtime:

Reviewer: ```suggestion
In order for your application to run on Cloudflare Pages, it needs to be set to use the Edge Runtime. Make sure that all the files in your application containing server-side code (e.g. any API Routes and any pages which use `getServerSideProps`) export a `config` object specifying the use of the Edge Runtime:
```

```ts
export const config = {

> 	runtime: 'edge',

Reviewer: This works fine for Edge API Routes, but I think will log a warning for an SSR pages which prompt users to continue using `experimental-edge`. This was true at least a little while ago. It might have now changed, rendering this comment moot. Feel free to disregard if I'm operating on outdated info.

Author: I think that `edge` is fine everywhere but if you don't provide it, our own warnings prompt you to use `experimental-edge` (instead of `edge`) and if you provide do `experimental-edge` you get a warning from the Vercel CLI.

I didn't want to add that change in this PR to keep the scope of this one to be just documentation, so I've updated our warning here instead: https://github.com/cloudflare/next-on-pages/pull/127

};
```

> Make also sure that your application is not using [unsupported APIs](https://nextjs.org/docs/api-reference/edge-runtime#unsupported-apis) and its API routes are defined as [Edge API Routes](https://nextjs.org/docs/api-routes/edge-api-routes).

Reviewer: ```suggestion
Additionally, ensure that your application is not using any [unsupported APIs](https://nextjs.org/docs/api-reference/edge-runtime#unsupported-apis) and that its API routes are defined as [Edge API Routes](https://nextjs.org/docs/api-routes/edge-api-routes).
```

```sh
cd my-next-app
```

1. Add a `NODE_VERSION` environment variable set to `16` or greater (`18` is not supported yet, See [Build Image Update Discussion](https://github.com/cloudflare/pages-build-image/discussions/1).
### 2. Configure the application to use the Edge Runtime

1. In the Pages project **Settings** > **Functions** > **Compatibility Flags**, add the `nodejs_compat` and ensure the **Compatibility Date** is set to at least `2022-11-30`.
In order for the application to run on Cloudflare Pages it needs to be set to run on the edge runtime, in order to do so make sure that all the files in your application containing server side code (as your api routes and pages using `getServerSideProps`) export a config object specifying the use of the edge runtime:

1. The project should now be ready to deploy. Create a new deployment.
```ts
export const config = {
	runtime: 'edge',
};
```

## `@cloudflare/next-on-pages` CLI
Make also sure that your application is not using [unsupported APIs](https://nextjs.org/docs/api-reference/edge-runtime#unsupported-apis) and its API routes are defined as [Edge API Routes](https://nextjs.org/docs/api-routes/edge-api-routes).

```

> ⚡️ @cloudflare/next-to-pages CLI

Author: I don't think showing the help message is very valuable since the developer can easily see it using `--help`, showing the message here also has the risk of this getting out of sync with the current help message, so I changed this section to: `If you provide the '--help' flag a useful help message will be displayed showing you the various options the CLI offers.`

> For example if you've created a Next.js application and opted out of both the `src` and `app` directory options, the only file you need to modify is the `pages/api/hello.ts` one.

Reviewer: ```suggestion
For example, if you've created a Next.js application with `create-next-app` and opted out of both the `src` and `app` directory options, the only file you need to modify is `pages/api/hello.ts`.
```

> You can easily deploy to Cloudflare Pages via the [automatic git integration](https://developers.cloudflare.com/pages/platform/git-integration/), to do so, start by committing and pushing your application's code to a Github/GitLab repository.

Reviewer: ```suggestion
You can easily deploy to Cloudflare Pages via the [automatic git integration](https://developers.cloudflare.com/pages/platform/git-integration/). To do so, start by committing and pushing your application's code to a GitHub/GitLab repository.
```

> Next in the [Cloudflare Dashboard](dash.cloudflare.com) to create a new Pages project:

Reviewer: ```suggestion
Next, in the [Cloudflare Dashboard](https://dash.cloudflare.com/?to=/:account/pages), create a new Pages project:
```

>   | Build command | `npx @cloudflare/next-on-pages --experimental-minify` |

Reviewer: Do we want to start suggesting that people pin their `@cloudflare/next-on-pages` version? Or keep it to latest (esp. when in 0.X.X releases)?

Author: I am not sure.... 😕

do you think we should pin it? (since we don't pin next itself it kind of feels natural to me also not to pin `@cloudflare/next-on-pages`?)

Reviewer: I think we should pin when get to 1.0.0, but probably not bother before that.

Author: sure, sounds good to me :slightly_smiling_face::+1:

> > If you don't want to set up a git repository, you can build your `_worker.js` file (as indicated in [Local Development](#local-development)) and publish your application manually via the [wrangler's pages publish command](https://developers.cloudflare.com/workers/wrangler/commands/#publish-1) instead (but you'll still need to set the `NODE_VERSION` environment variable and the `nodejs_compat` flag for your project in the Cloudflare dashboard).

Reviewer: ```suggestion
> If you don't want to set up a git repository, you can build your `_worker.js` file (as indicated in [Local Development](#local-development)) and publish your application manually via the [wrangler's pages publish command](https://developers.cloudflare.com/workers/wrangler/commands/#publish-1) instead (but you'll still need to set the `nodejs_compat` flag for your project in the Cloudflare dashboard).
```

> This command will build your Next.js application and bundle a `_worker.js` file for you to use.

Reviewer: I don't think most people know about (or care about) `_worker.js`. Instead, we should frame this as:

````suggestion
This command will build your Next.js application and produce a `.vercel/output/static` directory which you can then use with Wrangler:

```sh
npx wrangler pages dev .vercel/output/static --compatibility-flag=nodejs_compat
```
````

> If you provide the `--help` flag a useful help message will be displayed showing you the various options the CLI offers.

Reviewer: ```suggestion
Running `npx @cloudflare/next-on-pages --help` will display a useful help message which will detail the various additional options the CLI offers.
```

> If you want to work on your Next.js application while using next-on-pages, run the CLI in watch mode via:

Reviewer: ```suggestion
If you want to work on your Next.js application while using next-on-pages, run the CLI in watch mode with:
```

```sh
npx @cloudflare/next-on-pages --watch
```

Then in a separate terminal run:

```sh
npx wrangler pages dev .vercel/output/static --compatibility-flag=nodejs_compat
```

> to view the application and having it refresh when you make changes to your code.

Reviewer: ```suggestion
```

### Install next-on-pages and vercel (optional)

> To speed up local development (especially the refresh speed in watch mode) you can install next-on-pages and `vercel` as dev dependencies of your project, note that this is optional and if you decide to install only one of them or neither everything will still run correctly anyways:

Reviewer: ```suggestion
To speed up local development (especially the refresh speed when running in watch mode) you can optionally choose to install `@cloudflare/next-on-pages` and `vercel` as dev dependencies of your project:
```

```sh
npm install -D @cloudflare/next-on-pages vercel
```

## Examples

To see some examples on how to use Next.js features with next-on-pages see the [Examples document](./docs/examples.md).

## Contributing

> If you want to contribute to the project please refer to the [Contributing document](./docs/contributing.md).

Reviewer: ```suggestion
If you want to contribute to this project please refer to the [Contributing document](./docs/contributing.md).
```

## References

Extra references you might be interested in:

- [Blog post](https://blog.cloudflare.com/next-on-pages)

>   The original blog post introducing next-on-pages (24/10/2022), it goes into details on how the package came to be and provides details on how it works.

Reviewer: ```suggestion
  The original blog post introducing next-on-pages (24/10/2022), it goes into details on the inspiration for this package and provides some details on how it works.
```

- [Cloudflare Next.js Guide](https://developers.cloudflare.com/pages/framework-guides/deploy-a-nextjs-site/)

>   Cloudflare guide on how to create and deploy a Next.js application. The application can be either static (and deployed as simple static assets) or dynamic using the edge runtime (via next-on-pages).

Reviewer: ```suggestion
  Cloudflare guide on how to create and deploy a Next.js application. The application can be either static (and deployed as simple static assets) or dynamic using the Edge Runtime (powered by `@cloudflare/next-on-pages`).
```

Author: Mh... next-on-pages isn't powering the edge runtime, is it? 🤔
